### ISSUE TITLE
Reachability analysis does not understand control flow exceptions (StopIteration)

### ISSUE THREAD

**Issue by mikeshardmind** [microsoft/pyright#9717]
```python
def _consumer[**P, Y](
    queue: Queue[Y],  # This is a custom queue class, but not interacting with this issue
    cancel_future: cf.Future[None],  # cf = concurrent.futures
    f: Callable[P, Generator[Y]],
    *args: P.args,
    **kwargs: P.kwargs,
) -> None:
    gen = f(*args, **kwargs)

    while True:
        val = next(gen)  # This raises StopIteration, exiting the loop.
        queue.sync_put(val)
      
        if cancel_future.cancelled():
            gen.throw(cf.CancelledError)  # This is being used due to this being an N:M concurrency tool to wrap generators as coroutines

    gen.close()  # pyright incorrectly says that this,and other cleanup is unreachable, and does not type check it.
```

If your code relies on symbols that are imported from a third-party library, include the associated import statements and specify which versions of those libraries you have installed.

**VS Code extension or command-line**
As a vs code extension

-   Language Server version: 2024.12.1
-   OS and version: win32 x64
-   python.analysis.indexing: true
-   python.analysis.typeCheckingMode: strict

**Comment by erictraut** [comment]
Here's a minimal code sample that reflects the behavior of the above code.

```python
def func1():
    while True:
        func2()
    print("here")

def func2():
    raise StopIteration()

func1()
```

If you run this code, you can see that the `print` statement is never executed. An exception raised by `func2` (or by `next` in your original code sample) will not cause control to fall out of the `while True` loop. The raised exception will propagate.

Pyright is working correctly here, so I don't consider this a bug. Not surprisingly, mypy also reports this code as unreachable.

**Comment by mikeshardmind** [comment]
Sorry about the noise on this one, I forwarded something internal without noticing the more meaningful change that broke the contract here.
